gensim/models/keyedvectors.py

```diff
@@ -1520,45 +1520,52 @@ def save_word2vec_format(
             (in case word vectors are appended with document vectors afterwards).
         write_header : bool, optional
             If False, don't write the 1st line declaring the count of vectors and dimensions.
-        FIXME: doc prefix, append, sort_attr
+        prefix : str, optional
+            String to prepend in front of each stored word. Default = no prefix.
+        append : bool, optional
+            If set, open `fname` in `ab` mode instead of the default `wb` mode.
+        sort_attr : str, optional
+            Sort the output vectors in descending order of this attribute. Default: most frequent keys first.
+
         """
         if total_vec is None:
             total_vec = len(self.index_to_key)
         mode = 'wb' if not append else 'ab'
-        if 'count' in self.expandos:
-            # if frequency-info available, store in most-to-least-frequent order
-            store_order_vocab_keys = sorted(self.key_to_index.keys(), key=lambda k: -self.get_vecattr(k, sort_attr))
-        else:
-            store_order_vocab_keys = self.index_to_key
+        if sort_attr not in self.expandos:
+            raise ValueError(f"attribute {sort_attr} not present in {self}")
+        store_order_vocab_keys = sorted(self.key_to_index.keys(), key=lambda k: -self.get_vecattr(k, sort_attr))
 
         if fvocab is not None:
             logger.info("storing vocabulary in %s", fvocab)
             with utils.open(fvocab, mode) as vout:
                 for word in store_order_vocab_keys:
-                    vout.write(utils.to_utf8("%s%s %s\n" % (prefix, word, self.get_vecattr(word, sort_attr))))
+                    vout.write(f"{prefix}{word} {self.get_vecattr(word, sort_attr)}\n")
 
         logger.info("storing %sx%s projection weights into %s", total_vec, self.vector_size, fname)
         assert (len(self.index_to_key), self.vector_size) == self.vectors.shape
 
-        # after (possibly-empty) initial range of int-only keys,
-        # store in sorted order: most frequent keys at the top
+        # After (possibly-empty) initial range of int-only keys,
+        # store in sorted order: most frequent keys at the top.
+        # TODO: is this still relevant in 4.0? What is this about?
```

Comment:
@gojomo is this still relevant? What comment goes here?

Reply:
This is set up so that *if* this instance has taken advantage of (any number of initial) literal int index keys to avoid having a larger `key_to_index` dict, those indexes are preserved (because changing them would be disruptive), before the rest of the key-looks-up-index indirectly keys (which don't need to preserve exact indexes across saves). 

(This optimized alternative was added for `Doc2Vec`, where doc-tags generally *don't* have a frequency order, and often can be just natural-ordinal numbers. In some plausible-but-extreme scenarios, like tens/hundreds-of-millions of doctags, this could save a notable amount of memory.)

Reply:
So it's a just backward compatibility thing? No possibility of new models (created in 4.0) being affected?

If so, I'll move it to some migration / load code path. That "optimized Doc2Vec code path with ints" I thought (hoped) we got rid of during the refactor. Way more trouble than worth.

Reply:
It's not just backward-compatibility - this optimization is still active & available to anyone who uses only plain ascending int doc tags in `Doc2Vec`. 

The original motivation was I had a target use where there were tens-of-millions of docs, and the RAM savings from avoiding the strings and the full dict was noticeable given the 16GB or 32GB machine I had to work with. (It's possible that the same sort of trick might be helpful for some outlier Word2Vec uses, too – perhaps in bioinformatics or log/network-analysis – but I haven't actually tried it.)

The code's been simplified somewhat, in that `.index_to_key` will always have entries for these direct-int slots. (There had been a further complication in earlier code trying to minimize or eliminate the list-of-keys - but the list overhead for ints seems much smaller than the dict overhead for strings.)

Reply:
Needs way better documentation in the code then. I thought the KeyedVectors keys are keys, and all code (that I saw or wrote) relies on that.

What are the actual invariants?

I'm strongly in favour of discontinuing this optimization in 4.0.